src/v/cluster/cluster_link/table.cc

```diff
@@ -172,7 +240,22 @@ void table::run_callbacks(id_t id) {
     }
 }
 
-cluster::errc table::upsert_link(id_t id, metadata meta) {
+cluster::cluster_link::errc table::upsert_link(id_t id, metadata meta) {
+    for (const auto& t : meta.state.mirror_topics) {
+        auto link_id = find_id_by_topic(t.first);
+        if (link_id && link_id.value() != id) {
+            vlog(
+              cluster::clusterlog.info,
+              "Unable to upsert link {} ({}) as topic {} is already registered "
+              "to link {}",
+              meta.name,
+              id,
+              t.first,
+              link_id.value());
+            return errc::topic_being_mirrored_by_other_link;
+        }
+        _topic_name_index.emplace(t.first, id);
```

Comment:
seems like you might be able to avoid the double container lookup with something like
```
if (!link_id)
    emplace
elif link_id.value != id
    return error
```
but that's based on the expected behavior here. asserting on the return value of emplace is useful to document expected behavior when the container may already contain the key.